core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -76,6 +95,9 @@ private boolean commit0(Substream winningSubstream) {
       state = state.committed(winningSubstream);
     }
 
+    // subtract the share of this RPC from channelBufferUsed.
+    channelBufferUsed.addAndGet(-perRpcBufferUsedUpdater.get(this));
```

Comment:
Can you explain what is racy?

Reply:
thread1:
        if (perRpcBufferUsedUpdater
            .compareAndSet(RetriableStream.this, savedPerRpcBufferUsed, bufferNeeded)) {
thread2:
        channelBufferUsed.addAndGet(-perRpcBufferUsedUpdater.get(this));
thread1:
          long savedChannelBufferUsed =
              channelBufferUsed.addAndGet(bufferNeeded - savedPerRpcBufferUsed);

Reply:
thread1 and 2 are same substream?

Reply:
I think the final state is correct and consistent, although between step 2 and step 3 the channel quote is lower than the actual quote, but it is only for a short moment. After all the buffer limit is only a rough estimate, not the actual physical memory size.

Reply:
Which streams are involved doesn't matter. The problem is that we "leaked" buffer in channelBufferUsed. thread1 increased channelBufferUsed and it will never be reversed.

Reply:
I agree to use the lock rather than atomic `perRpcBufferUsedUpdater` which takes too much effort and still cannot guarantee thread safety.